libsolidity/lsp/Utils.h

```diff
@@ -0,0 +1,82 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#pragma once
+
+#include <liblangutil/SourceLocation.h>
+
+#include <libsolidity/ast/ASTForward.h>
+
+#include <libsolutil/JSON.h>
+
+#include <optional>
+#include <vector>
+
+#if !defined(NDEBUG)
+#include <fstream>
+#define lspDebug(message) (std::ofstream("/tmp/solc.log", std::ios::app) << (message) << std::endl)
+#else
+#define lspDebug(message) do {} while (0)
+#endif
+
+namespace solidity::langutil
+{
+class CharStreamProvider;
+}
+
+namespace solidity::lsp
+{
+
+class FileRepository;
+
+std::optional<langutil::LineColumn> parseLineColumn(Json::Value const& _lineColumn);
+Json::Value toJson(langutil::LineColumn _pos);
+Json::Value toJsonRange(langutil::LineColumn const& _start, langutil::LineColumn const& _end);
+
+/// @returns the source location given a source unit name and an LSP Range object,
+/// or nullopt on failure.
+std::optional<langutil::SourceLocation> parsePosition(
+	FileRepository const& _fileRepository,
+	std::string const& _sourceUnitName,
+	Json::Value const& _position
+);
+
+/// @returns the source location given a source unit name and an LSP Range object,
+/// or nullopt on failure.
+std::optional<langutil::SourceLocation> parseRange(
+	FileRepository const& _fileRepository,
+	std::string const& _sourceUnitName,
+	Json::Value const& _range
+);
+
+/// Extracts all resolved annotations of the given expression AST node.
+///
+/// This may for example by the type declaration of an identifier,
+/// or the type declaration of a structured member identifier.
+///
+/// The returned vector is empty if the expression was not resolved (maybe due to syntax/semantic error),
+/// or one if resolved.
+/// In the future, it may be possible to return more than one type declaration,
+/// if the type resolver did find more than one possible candidate declaration.
```

Comment:
Wouldn't it be better to return  nothing instead of multiple candidates? If we return multiple candidates, the result will be wrong, won't it?

Reply:
I think so.